server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryTest.java

```diff
@@ -0,0 +1,125 @@
+package org.apache.james.mailrepository;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertAll;
+
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.StringJoiner;
+
+import javax.mail.MessagingException;
+
+import org.apache.commons.lang.RandomStringUtils;
+import org.apache.james.core.MailAddress;
+import org.apache.james.mailrepository.api.MailRepository;
+import org.apache.james.mime4j.io.InputStreams;
+import org.apache.james.server.core.MailImpl;
+import org.apache.mailet.Mail;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+
+import com.google.common.collect.ImmutableList;
+
+public interface MailRepositoryTest {
+
+    String TEST_ATTRIBUTE = "testAttribute";
+
+    default  Mail createMail(String name) throws MessagingException {
+        InputStream mailContent = generateMailContent();
+        List<MailAddress> recipients = ImmutableList
+            .of(new MailAddress("rec1@domain.com"),
+                new MailAddress("rec2@domain.com"));
+        MailAddress sender = new MailAddress("sender@domain.com");
+        Mail mail = new MailImpl(name, sender, recipients, mailContent);
+        mail.setAttribute(TEST_ATTRIBUTE, "testValue");
+        return mail;
+    }
+
+    default InputStream generateMailContent() {
+        String headers = new StringJoiner("\r\n")
+            .add("Subject: test")
+            .add("Content-Type: text/plain")
+            .toString();
+        String headerBodySeparator = "\r\n\r\n";
+        String body = "original body" + RandomStringUtils.randomAlphabetic(10);
```

Comment:
why introducing randomness?
I guess as this is just porting a test you miss a remove to clearly show that you are just moving some tests, no?

Reply:
To be honest, I've read "Found more than one message after storing 2 times the SAME message: it MUST update the previous" in the Abstract version, but the assertion was not testing that. So I wrote a test about that to understand what it probably expected to test back then.
The problem is : you do make mails with different contents, you write a new method that takes a body ? I don't like that very much, method overloading is painful. Modify Mail then call store with the modified version ? I don't trust Mail implementation enough for that. Write a Mail.Builder ? I hesitate for some time but Hell, all these classes are a mess, I'll take forever ):

So sadly, I've chose the quickest solution for now. I'm pretty sure I'll have to challenge that before the end of the ticket.

Do you think we can merge that for now ?

Reply:
How about making FakeMail serializable; and relying on FakeMail? Some other fixes along the way, but I succeeded writing all MailQueue Tests with FakeMail.

Reply:
It's one more point of frustration : Mail is just a datastructure, it doesn't make sense to have an interface and several impl.
I digged that yesterday but I'd rather merge FakeMail and MailImpl than use FakeMail more. I'm really not confidence use FakeMail or MailImpl in tests gives the same result today

Reply:
let's go just adding a method that take a content in parameter to build different mails, I find it cleaner than this randomness